AsyncDisplayKit/Private/ASDisplayNode+UIViewBridge.mm

```diff
@@ -224,47 +246,31 @@ - (void)setFrame:(CGRect)rect
     // Checking if the transform is identity is expensive, so disable when unnecessary. We have assertions on in Release, so DEBUG is the only way I know of.
     ASDisplayNodeAssert(CATransform3DIsIdentity(self.transform), @"-[ASDisplayNode setFrame:] - self.transform must be identity in order to set the frame property.  (From Apple's UIView documentation: If the transform property is not the identity transform, the value of this property is undefined and therefore should be ignored.)");
 #endif
+    _view.frame = rect;
+  } else if (!nodeLoaded || (isMainThread && !setFrameDirectly)) {
+    /**
+     * Sets a new frame to this node by changing its bounds and position. This method can be safely called even if
+     * the transform is a non-identity transform, because bounds and position can be set instead of frame.
+     * This is NOT called for synchronous nodes (wrapping regular views), which may rely on a [UIView setFrame:] call.
+     * A notable example of the latter is UITableView, which won't resize its internal container if only layer bounds are set.
+     */
+    CGRect bounds;
+    CGPoint position;
+    ASBoundsAndPositionForFrame(rect, self.bounds.origin, self.anchorPoint, &bounds, &position);
 
-    _setToViewOnly(frame, rect);
-  } else {
-    // This is by far the common case / hot path.
-    [self __setSafeFrame:rect];
-  }
-}
-
-/**
- * Sets a new frame to this node by changing its bounds and position. This method can be safely called even if
- * the transform is a non-identity transform, because bounds and position can be set instead of frame.
- * This is NOT called for synchronous nodes (wrapping regular views), which may rely on a [UIView setFrame:] call.
- * A notable example of the latter is UITableView, which won't resize its internal container if only layer bounds are set.
- */
-- (void)__setSafeFrame:(CGRect)rect
-{
-  ASDisplayNodeAssertThreadAffinity(self);
-  ASDN::MutexLocker l(_propertyLock);
-  
-  BOOL useLayer = (_layer && ASDisplayNodeThreadIsMain());
-  
-  CGPoint origin      = (useLayer ? _layer.bounds.origin : self.bounds.origin);
-  CGPoint anchorPoint = (useLayer ? _layer.anchorPoint   : self.anchorPoint);
-  
-  CGRect  bounds      = (CGRect){ origin, rect.size };
-  CGPoint position    = CGPointMake(rect.origin.x + rect.size.width * anchorPoint.x,
-                                    rect.origin.y + rect.size.height * anchorPoint.y);
-  
-  if (useLayer) {
-    _layer.bounds = bounds;
-    _layer.position = position;
-  } else {
     self.bounds = bounds;
     self.position = position;
+  } else if (nodeLoaded && !isMainThread) {
+    if (!_pendingViewState.hasChanges) {
+      [ASPendingStateController.sharedInstance registerNode:self];
```

Comment:
Eh not really. Currently the only reusable presence of `registerNode:` is in `ASDisplayNodeShouldApplyBridgedWriteToView` which is a very fast package of [check node, check thread, dirty if needed, return if should set to view]. In setFrame: we need to use some of those intermediate results. We could call it here, and it wouldn't result in any more method calls but it would result in another ASDisplayNodeThreadIsMain and another node-loadedness check. Doesn't seem worth it to me, let me know if you disagree.